File: 2021/9/python/main.py
import numpy as np
from scipy.ndimage.interpolation import shift

with open('../input.txt') as f:
    lines = f.readlines()

lines = [l.strip() for l in lines]
a = np.array([[int(i) for i in l] for l in lines])

# Part 1

def is_minima(i, j):
    n = a[i,j]

    if i > 0 and a[i-1,j] <= n:
        return False
    if j > 0 and a[i,j-1] <= n:
        return False
    if i < (len(a) - 1) and a[i+1,j] <= n:
        return False
    if j < (len(a) - 1) and a[i,j+1] <= n:
        return False

    return True

risk_level = 0

for i in range(len(a)):
    for j in range(len(a)):
        if is_minima(i, j):
            #print(j, a[0,j])
            risk_level += 1 + a[i,j]

print(risk_level)

# Part 1 v2

minima = a[(
    (a < shift(a, (0, 1), cval=10)) & 
    (a < shift(a, (1, 0), cval=10)) &
    (a < shift(a, (0, -1), cval=10)) & 
    (a < shift(a, (-1, 0), cval=10))
)]

print(minima.sum() + len(minima))

# Part 2

b = a.copy()
sizes = []

def size_of_region(i, j):
    if b[i, j] in [-1, 9]:
        return 0

    s = 1
    b[i,j] = -1

    if i > 0:
        s += size_of_region(i-1, j)
    if j > 0:
        s += size_of_region(i, j-1)
    if i < (len(a) - 1):
        s += size_of_region(i+1, j)
    if j < (len(a) - 1):
        s += size_of_region(i, j+1)

    return s


for i in range(len(a)):
    for j in range(len(a)):
        s = size_of_region(i,j)
        if s:
            sizes.append(s)

print(np.prod(sorted(sizes)[-3:]))

File: 2021/9/rust/src/main.rs
use std::cmp::Reverse;
use std::error::Error;
use std::fs::File;
use std::io::{BufRead, BufReader};

fn is_minima(a: &Vec<Vec<i32>>, i: usize, j: usize) -> bool {
    let n = a[i][j];

    if i > 0 && a[i - 1][j] <= n {
        return false;
    }
    if j > 0 && a[i][j - 1] <= n {
        return false;
    }
    if i < (a.len() - 1) && a[i + 1][j] <= n {
        return false;
    }
    if j < (a.len() - 1) && a[i][j + 1] <= n {
        return false;
    }

    true
}

fn size_of_region(a: &mut Vec<Vec<i32>>, i: usize, j: usize) -> i32 {
    match a[i][j] {
        -1 | 9 => 0,
        _ => {
            let mut s = 1;
            a[i][j] = -1;

            if i > 0 {
                s += size_of_region(a, i - 1, j);
            }
            if j > 0 {
                s += size_of_region(a, i, j - 1);
            }
            if i < (a.len() - 1) {
                s += size_of_region(a, i + 1, j);
            }
            if j < (a.len() - 1) {
                s += size_of_region(a, i, j + 1);
            }

            s
        }
    }
}

fn main() -> Result<(), Box<dyn Error>> {
    let file = File::open("../input.txt")?;
    let reader = BufReader::new(file);

    let mut a: Vec<Vec<i32>> = reader
        .lines()
        .map(|line| {
            line.unwrap()
                .trim()
                .chars()
                .map(|c| i32::try_from(c.to_digit(10).unwrap()).unwrap())
                .collect()
        })
        .collect();

    let mut risk_level: i32 = 0;
    for i in 0..a.len() {
        for j in 0..a.len() {
            if is_minima(&a, i, j) {
                risk_level += 1 + a[i][j];
            }
        }
    }

    println!("Part 1: {}", risk_level);

    let mut sizes: Vec<i32> = Vec::new();

    for i in 0..a.len() {
        for j in 0..a.len() {
            let s = size_of_region(&mut a, i, j);
            if s > 0 {
                sizes.push(s);
            }
        }
    }

    sizes.sort_by_key(|s| Reverse(*s));

    println!("Part 2: {}", sizes[0] * sizes[1] * sizes[2]);

    Ok(())
}
